apps/emqx_dashboard/src/emqx_dashboard_swagger.erl

```diff
@@ -0,0 +1,228 @@
+components([], SpecAcc, SubRefAcc) -> components(SubRefAcc, SpecAcc, []);
+components([{Module, Field} | Refs], SpecAcc, SubRefsAcc) ->
+    Props = apply(Module, fields, [Field]),
+    case is_list(Props) of
+        true -> %% object
+            {Object, SubRefs} = parse_object(maps:from_list(Props)),
+            NewSpecAcc = SpecAcc#{?REF(Module, Field) => Object},
+            components(Refs, NewSpecAcc, SubRefs ++ SubRefsAcc);
+        false -> %% schema
+            Init0 = maps:without([type, desc], Props),
+            Init = trans_desc(Props, Init0),
+            {Spec, SubRef} = hocon_schema_to_spec(maps:get(type, Props)),
+            NewSpecAcc = SpecAcc#{?REF(Module, Field) => maps:merge(Spec, Init)},
+            components(Refs, NewSpecAcc, SubRef ++ SubRefsAcc)
+    end.
+
+hocon_schema_to_spec(?R_REF(Module, StructName)) ->
+    {#{<<"$ref">> => ?TO_COMPONENTS(Module, StructName)}, [{Module, StructName}]};
+hocon_schema_to_spec(Type) when ?IS_TYPEREFL(Type) ->
+    Match = #{
+        boolean() => {boolean, true},
+        binary() => {string, <<"binary example">>},
+        float() => {number, 3.14159},
+        integer() => {integer, 100},
+        number() => {number, 42},
+        string() => {string, <<"string example">>},
+        atom() => {string, atom},
+        duration() => {string, "12m"},
+        duration_s() => {string, "1h"},
+        duration_ms() => {string, "32s"},
+        percent() => {number, "12%"},
+        file() => {string, "/var/log/emqx/"},
+        ip_port() => {string, "127.0.0.1:80"}
+    },
+    case maps:find(Type, Match) of
+        {ok, {T, Example}} ->
+            {#{type => T, example => Example}, []};
+        error ->
+            [MinStr, MaxStr] = string:split(typerefl:print(Type), ".."),
+            {Min, []} = string:to_integer(MinStr),
+            {Max, []} = string:to_integer(MaxStr),
+            {#{type => integer, example => Min, minimum => Min, maximum => Max}, []}
+    end;
+hocon_schema_to_spec(?ARRAY(Item)) ->
+    {Schema, Refs} = hocon_schema_to_spec(Item),
+    {#{type => array, items => Schema}, Refs};
+hocon_schema_to_spec(?ENUM(Items)) ->
+    {#{type => string, enum => Items}, []};
+hocon_schema_to_spec(?UNION(Types)) ->
+    {OneOf, Refs} = lists:foldl(fun(Type, {Acc, RefsAcc}) ->
+        {Schema, SubRefs} = hocon_schema_to_spec(Type),
+        {[Schema | Acc], SubRefs ++ RefsAcc}
+                                end, {[], []}, Types),
+    {#{<<"oneOf">> => OneOf}, Refs};
+hocon_schema_to_spec(Atom) when is_atom(Atom) ->
+    {#{type => string, enum => [Atom]}, []}.
+
+to_bin(List) when is_list(List) -> list_to_binary(List);
+to_bin(Atom) when is_atom(Atom) -> atom_to_binary(Atom, utf8);
+to_bin(X) -> X.
+
+parse_object(PropList) when PropList =:= #{} -> {[], []};
+parse_object(PropList) ->
+    {Props, Required, Refs} =
+        maps:fold(fun(Name, Hocon, {Acc, RequiredAcc, RefsAcc}) ->
+            NameBin = to_bin(Name),
```

Comment:
we may need to prepend the name with `hocon_schema:namespace(SchemaModule)` (use `:` as delimiter).
note: it may return `undefined`.